### PR TITLE
Added rag over code example

### PR DESCRIPTION
**Pull request by bracesproul**
Context: in the runnable branch I'm checking if chat messages are > 1 or < 1 since `ConversationSummaryMemory` appears to insert a default empty system message on initialization.

### CODE REVIEW

File: docs/docs/use_cases/rag/_category_.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+label: 'RAG'
```
Comment (jacoblee93) [inline docs/docs/use_cases/rag/_category_.yml:1]: Change to `Retrieval-augmented generation (RAG)`?

File: docs/docs/use_cases/rag/code_understanding.mdx [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+- Q&A over the code base to understand how it works
+- Using LLMs for suggesting refactors or improvements
+- Using LLMs for documenting the code
+
+![RAG over code](/img/rag_code_diagram.png)
+
+## Overview
+
+The pipeline for QA over code follows the [steps we do for document question answering](http://localhost:3000/docs/use_cases/question_answering/), with some differences:
```
Comment (jacoblee93) [inline docs/docs/use_cases/rag/code_understanding.mdx:15]: Remove `localhost:3000` link, use just `/docs/`

File: docs/docs/use_cases/rag/code_understanding.mdx [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+import { BaseMessage } from "langchain/schema";
+import { StringOutputParser } from "langchain/schema/output_parser";
+```
+
+First we construct the combine documents chain. This chain takes in the question, chat history (from memory) and context (relevant documents). It then passes all three into the prompt which answers the users question.
+
+```typescript
+const combineDocumentsPrompt = ChatPromptTemplate.fromMessages([
+  [
+    "ai",
```
Comment (jacoblee93) [inline docs/docs/use_cases/rag/code_understanding.mdx:168]: This is the common paradigm:

```js
    const chatPrompt = ChatPromptTemplate.fromMessages([
      SystemMessagePromptTemplate.fromTemplate(
        "You are a helpful assistant who must always respond like a pirate"
      ),
      new MessagesPlaceholder("history"),
      HumanMessagePromptTemplate.fromTemplate("{input}"),
    ]);
```

This prompting pattern will work for some models but not others (e.g. Anthropic)

File: docs/docs/use_cases/rag/code_understanding.mdx [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+```typescript
+const question_generator_template = ChatPromptTemplate.fromMessages([
+  [
+    "ai",
+    `Given the following conversation about a codebase and a follow up question, rephrase the follow up question to be a standalone question.`,
+  ],
+  new MessagesPlaceholder("chat_history"),
+  [
+    "ai",
```
Comment (jacoblee93) [inline docs/docs/use_cases/rag/code_understanding.mdx:205]: See above

File: docs/docs/use_cases/rag/code_understanding.mdx [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+  {
+    question: (i: { question: string }) => {
+      return questionGeneratorChain.invoke(i);
+    },
+  },
+  combineDocumentsChain,
+]);
+```
+
+Finally, we'll patch it all together using a [`RunnableBranch`](/docs/expression_language/how_to/routing#using-a-runnablebranch).
```
Comment (jacoblee93) [inline docs/docs/use_cases/rag/code_understanding.mdx:242]: I think simpler to not include this? Or include it as an addendum at the end?

**Review by jacoblee93 — DISMISSED** [review]
x

File: docs/docs/use_cases/rag/code_understanding.mdx (hunk 5 above)
Comment (bracesproul) [reply docs/docs/use_cases/rag/code_understanding.mdx:242]: We don't necessarily **need** it, but in order to be able to ask follow ups we need the branch to check if there is chat history or not. Although we could omit it, and have the followup explicitly call `questionGeneratorAndCombineDocumentsChain`, it's a bad practice compared to using a branch and I don't think we should use it in our use cases.

Comment (jacoblee93) [reply docs/docs/use_cases/rag/code_understanding.mdx:242]: Just always have the rephrase step in the initial version?

Comment (bracesproul) [reply docs/docs/use_cases/rag/code_understanding.mdx:242]: Ah yeah that would work. Refactoring rn

Comment (bracesproul) [reply docs/docs/use_cases/rag/code_understanding.mdx]: Pushed

File: docs/docs/use_cases/rag/code_understanding.mdx [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+  new MessagesPlaceholder("chat_history"),
+  HumanMessagePromptTemplate.fromTemplate("Question: {question}"),
+]);
+```
+
+Next, we'll construct our chain.
+
+```typescript
+const conversationalQaChain = RunnableSequence.from([
+  {
```
Comment (jacoblee93) [inline docs/docs/use_cases/rag/code_understanding.mdx:208]: I would say breaking these up into smaller logical sequences might help with readability?

File: docs/docs/use_cases/rag/code_understanding.mdx [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+Now we can construct our main sequence of chains. We're going to be building `ConversationalRetrievalChain` using Expression Language.
+
+```typescript
+import {
+  ChatPromptTemplate,
+  MessagesPlaceholder,
+  AIMessagePromptTemplate,
+  HumanMessagePromptTemplate,
+} from "langchain/prompts";
+import { RunnableBranch, RunnableSequence } from "langchain/schema/runnable";
```
Comment (jacoblee93) [inline docs/docs/use_cases/rag/code_understanding.mdx:162]: Doesn't look like you're using `RunnableBranch` anymore?

**Review by jacoblee93 — CHANGES_REQUESTED** [review]
Seems reasonable, two small notes

File: docs/docs/use_cases/rag/code_understanding.mdx (hunk 6 above)
Comment (bracesproul) [reply docs/docs/use_cases/rag/code_understanding.mdx:212]: Good idea, split into two, one for combining docs and one for the conversational part

**Review by jacoblee93 — APPROVED** [review]

**Pull request merged**